drivers/sensor/bma4xx/bma4xx.c

```diff
@@ -0,0 +1,687 @@
+/* Bosch BMA4xx 3-axis accelerometer driver
+ *
+ * Copyright (c) 2023 Google LLC
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#define DT_DRV_COMPAT bosch_bma4xx
+
+#include <zephyr/init.h>
+#include <zephyr/sys/byteorder.h>
+#include <zephyr/sys/__assert.h>
+#include <zephyr/logging/log.h>
+#include <zephyr/pm/device.h>
+
+LOG_MODULE_REGISTER(bma4xx, CONFIG_SENSOR_LOG_LEVEL);
+#include "bma4xx.h"
+
+#ifdef CONFIG_PM_DEVICE
+static int bma4xx_pm_action(const struct device *dev, enum pm_device_action action)
+{
+	return -ENOTSUP;
+}
+#endif /* CONFIG_PM_DEVICE */
+
+/**
+ * Perform a read-modify-write on a specific register.
+ */
+static int bma4xx_reg_update(const struct device *dev, uint8_t reg, uint8_t mask, uint8_t new_value)
+{
+	uint8_t temp, orig_value;
+	int status;
+	struct bma4xx_data *bma4xx = dev->data;
+
+	status = bma4xx->hw_ops->read_reg(dev, reg, &temp);
+	if (status) {
+		LOG_ERR("Could not read initial value: %d", status);
+		return status;
+	}
+	orig_value = temp;
+
+	/* Mask out and overwrite bits */
+	temp &= ~mask;
+	temp |= new_value & mask;
+
+	if (orig_value == temp) {
+		/* No write necessary */
+		return 0;
+	}
+
+	/* Write back */
+	status = bma4xx->hw_ops->write_reg(dev, reg, temp);
+	if (status) {
+		LOG_ERR("Could not write updated value: %d", status);
+		return status;
+	}
+
+	return 0;
+}
+
+/**
+ * Set the X, Y, or Z axis offsets.
+ */
+static int bma4xx_attr_set_offset(const struct device *dev, enum sensor_channel chan,
+				  const struct sensor_value *val)
+{
+	struct bma4xx_data *bma4xx = dev->data;
+	uint8_t reg_addr;
+
+	switch (chan) {
+	case SENSOR_CHAN_ACCEL_X:
+		reg_addr = BMA4XX_REG_OFFSET_0;
+		break;
+	case SENSOR_CHAN_ACCEL_Y:
+		reg_addr = BMA4XX_REG_OFFSET_1;
+		break;
+	case SENSOR_CHAN_ACCEL_Z:
+		reg_addr = BMA4XX_REG_OFFSET_2;
+		break;
```

Comment:
It's more likely we'll pass all 3 values via SENSOR_CHAN_ACCEL_XYZ in order to perform only 1 bus i/o. Can we support that path? If it's too hard I would prefer to support the XYZ code path over the 3 individuals

Reply:
Ok the `*val` pointer would lead to an array of three `struct sensor_value`s in that case?